features/configure-users/configure.d/configure-users

```diff
@@ -0,0 +1,180 @@
+  _USER_FILE="$(mktemp /tmp/cluster-customizer.user-list.XXXXXXXX)"
+  _GROUP_FILE="$(mktemp /tmp/cluster-customizer.group-list.XXXXXXXX)"
+  if [ -n "${cw_CLUSTER_CUSTOMIZER_bucket}" ]; then
+    _pull_files "${cw_CLUSTER_CUSTOMIZER_bucket}"
+  elif network_is_ec2; then
+    _pull_files "s3://alces-flight-$(network_ec2_hashed_account)"
+  else
+    echo "ERROR: Clusterware s3 bucket has not been set" >&2
+    exit 1
+  fi
+}
+
+_pull_files() {
+  local bucket out
+  bucket=$1
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/userlist.cfg $_USER_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate user list: $out" >&2
+    exit 1
+  fi
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/grouplist.cfg $_GROUP_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate group list: $out" >&2
+    exit 1
+  fi
+}
+
+# Reads the groups from the list
+_add_groups_from_list() {
+  local delim regex line
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    #Validates input
+    regex="^[a-z][-a-z0-9]*$delim[0-9]+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_group $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised group format: $line" 1>&2
+      exit 1
+    fi
+  done < $_GROUP_FILE
+}
+
+# Reads the users from the list
+_add_users_from_list() {
+  local delim regex line
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    # Validates input
+    regex="^[a-z][-a-z0-9]*($delim[0-9]+)+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_user $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised user format: $line" 1>&2
+      exit 1
+    fi
+  done < $_USER_FILE
+}
+
+# Adds group ($1 groupname, $2 gid)
+_add_group() {
+  local groupname gid out
+  groupname=$1
+  gid=$2
+
+  out=$(groupadd -g $gid $groupname 2>&1)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not add group $groupname, $out" >&2
+    _EXIT_CODE=1
+  fi
+}
+
+# Adds user ($1 username, $2 uid, $3 gid, $4+ secondary gids)
+_add_user() {
+  local username uid gid gid2 out
+  username=$1
+  uid=$2
+  gid=$3
+  shift 3
+
+  # Sets gid to be uid if not included
+  if [[ -z "$gid" ]]; then
+    gid=$uid
+  fi
+
+  # Adds the user (and creates primary group if it doesn't exists)
+  getent group $gid >/dev/null || _add_group $username $gid
+  out=$(useradd -u $uid -g $gid $username 2>&1)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not add user $username, $out" >&2
+    _EXIT_CODE=1
+  fi
+
+  # Adds user to additional groups
+  if [[ "$#" -ne 0 ]]; then
+    for gid2 in $@; do
+      out=$(usermod -aG $gid2 $username 2>&1)
+      if [[ "$?" -ne 0 ]]; then
+        echo "ERROR: Could not add $username (user) to $@ (group), $out" >&2
+        _EXIT_CODE=1
```

Comment:
As the whole thing is failing at this point, maybe fail fast and do `return 1` after this?